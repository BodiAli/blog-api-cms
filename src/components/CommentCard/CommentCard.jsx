import { formatDistanceToNow } from "date-fns";
import anonymousImage from "../../assets/images/anonymous.jpg";
import optionsIcon from "../../assets/images/dots-vertical.svg";
import deleteIcon from "../../assets/images/delete.svg";
import styles from "./CommentCard.module.css";

export default function CommentCard({ comment, onDelete }) {
  return (
    <div className={styles.card}>
      <div className={styles.container1}>
        <div className={styles.commentUserInfo}>
          <img
            src={comment.User.Profile.profileImgUrl ? comment.User.Profile.profileImgUrl : anonymousImage}
            alt={`${comment.User.firstName} ${comment.User.lastName}'s profile picture`}
          />
          <p className={styles.name}>
            {comment.User.firstName} {comment.User.lastName}
          </p>
          <p className={styles.createdAt}>
            {formatDistanceToNow(comment.updatedAt, { addSuffix: true, includeSeconds: true })}
          </p>
        </div>
        <div className={styles.dropdown}>
          <button className={styles.dropBtn}>
            <img src={optionsIcon} alt="options" />
          </button>
          <div className={styles.dropdownContent}>
            <div onClick={() => onDelete(comment)} className={styles.folderDelete}>
              <img src={deleteIcon} alt="delete comment" />
              <button>Delete</button>
            </div>
          </div>
        </div>
      </div>
      <div className={styles.container2}>
        <p>{comment.content}</p>
      </div>
      <div className={styles.likeContainer}>
        <p className={styles.likes} data-testid="likes-count-comment">
          <strong>Likes:</strong> {comment.likes}
        </p>
      </div>
    </div>
  );
}
