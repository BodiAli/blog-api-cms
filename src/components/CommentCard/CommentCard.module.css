.card {
  padding: 0.5rem;
  display: grid;
  grid-template-columns: 1fr;
  gap: 2rem;
  border-radius: 5px;
  border-bottom: 1px solid rgba(0, 0, 0, 0.514);
  position: relative;
}

.container1 {
  display: grid;
  grid-template-columns: 1fr 1fr;
  align-items: center;
}

.container1 > .commentUserInfo {
  display: grid;
  grid-auto-flow: column;
  justify-content: start;
  align-items: center;
  gap: 0.5rem;
}

.container1 img {
  width: 60px;
  aspect-ratio: 1 / 1;
  object-fit: cover;
  border-radius: 50%;
}

.name {
  font-size: 1.1rem;
}

.createdAt {
  color: #0000009c;
}

.dropdown {
  justify-self: end;
}

.dropdown > button {
  background: none;
  border: none;
  cursor: pointer;
  transition: 150ms filter;
  &:hover {
    filter: invert(0.5);
  }
}

.dropdown img {
  width: 30px;
}

.dropdownContent {
  right: 0;
  display: none;
  position: absolute;
  background-color: #fff;
  border-radius: 5px;
  box-shadow: 0 8px 16px rgba(0, 0, 0, 0.2);
  width: 25%;
}

.dropdownContent > div {
  display: flex;
  align-items: center;
  gap: 10px;
  cursor: pointer;
  padding: 0.5rem;
  transition: 200ms background-color;
  &:hover {
    background-color: gainsboro;
  }
}

.dropdownContent button {
  flex: 1;
  text-align: start;
  background: none;
  border: none;
  font-size: 0.9rem;
}

.dropdown:hover .dropdownContent {
  display: block;
}

.likeContainer strong {
  font-weight: 500;
}

@media (max-width: 600px) {
  .container1 {
    grid-template-columns: auto auto;
  }

  .dropdownContent {
    width: 50%;
  }
}
