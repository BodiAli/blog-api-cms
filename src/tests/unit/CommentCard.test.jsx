import { render, screen } from "@testing-library/react";
import { createRoutesStub } from "react-router";
import userEvent from "@testing-library/user-event";
import CommentCard from "../../components/CommentCard/CommentCard";

vi.stubGlobal("fetch", vi.fn());

vi.mock("date-fns", async (importOriginal) => {
  const result = await importOriginal();
  return {
    ...result,
    formatDistanceToNow: vi.fn(() => "1 day ago"),
  };
});

const fakeComment = {
  id: "7397319e-58ef-4e42-bc94-aa8788db4090",
  userId: 1,
  postId: "57c20dfb-8135-42eb-8abe-0bffbaffd668",
  content: "Comment content",
  likes: 1,
  createdAt: "2025-05-03T07:51:49.273Z",
  updatedAt: "2025-05-04T06:19:03.092Z",
  User: {
    id: 1,
    firstName: "Bodi",
    lastName: "Ali",
    Profile: {
      profileImgUrl: "imageUrl",
    },
  },
};

const fakeOnDelete = vi.fn();

const Stub = createRoutesStub([
  {
    path: "/",
    Component: () => <CommentCard comment={fakeComment} onDelete={fakeOnDelete} />,
  },
]);

describe("CommentCard component", () => {
  test("Should render comment data", async () => {
    const { container } = render(<Stub />);

    expect(container).toMatchSnapshot();
  });

  test("Should call onDelete when delete button is clicked", async () => {
    const user = userEvent.setup();

    fetch.mockImplementationOnce(() => {
      return Promise.resolve({
        ok: true,
        status: 204,
      });
    });

    render(<Stub />);

    const deleteButton = screen.getByRole("button", { name: "Delete" });

    await user.click(deleteButton);

    expect(fakeOnDelete).toHaveBeenCalledWith(fakeComment);
  });
});
