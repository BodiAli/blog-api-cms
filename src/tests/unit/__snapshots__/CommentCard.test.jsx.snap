// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CommentCard component > Should render comment data 1`] = `
<div>
  <div
    class="_card_ec5f2a"
  >
    <div
      class="_container1_ec5f2a"
    >
      <div
        class="_commentUserInfo_ec5f2a"
      >
        <img
          alt="Bodi Ali's profile picture"
          src="imageUrl"
        />
        <p
          class="_name_ec5f2a"
        >
          Bodi
           
          Ali
        </p>
        <p
          class="_createdAt_ec5f2a"
        >
          1 day ago
        </p>
      </div>
      <div
        class="_dropdown_ec5f2a"
      >
        <button
          class="_dropBtn_ec5f2a"
        >
          <img
            alt="options"
            src="data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20viewBox='0%200%2024%2024'%3e%3cpath%20d='M12,16A2,2%200%200,1%2014,18A2,2%200%200,1%2012,20A2,2%200%200,1%2010,18A2,2%200%200,1%2012,16M12,10A2,2%200%200,1%2014,12A2,2%200%200,1%2012,14A2,2%200%200,1%2010,12A2,2%200%200,1%2012,10M12,4A2,2%200%200,1%2014,6A2,2%200%200,1%2012,8A2,2%200%200,1%2010,6A2,2%200%200,1%2012,4Z'%20/%3e%3c/svg%3e"
          />
        </button>
        <div
          class="_dropdownContent_ec5f2a"
        >
          <div
            class="_folderDelete_ec5f2a"
          >
            <img
              alt="delete comment"
              src="data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20viewBox='0%200%2024%2024'%3e%3cpath%20d='M9,3V4H4V6H5V19A2,2%200%200,0%207,21H17A2,2%200%200,0%2019,19V6H20V4H15V3H9M7,6H17V19H7V6M9,8V17H11V8H9M13,8V17H15V8H13Z'%20/%3e%3c/svg%3e"
            />
            <button>
              Delete
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="_container2_ec5f2a"
    >
      <p>
        Comment content
      </p>
    </div>
    <div
      class="_likeContainer_ec5f2a"
    >
      <p
        class="_likes_ec5f2a"
        data-testid="likes-count-comment"
      >
        <strong>
          Likes:
        </strong>
         
        1
      </p>
    </div>
  </div>
</div>
`;
